Add agent component overview document (#1544)
vdaas/vald: docs/overview/component/agent.md

# Vald Agent

Vald Agent is the core component of the Vald cluster.

Mainly, it consists of 1 or 2 small components, Vald Core and Vald Agent sidecar, depending on the intended use.

This page introduces the overview and features of Vald Agent.

## Responsibility

Vald Agent is responsible for:

> - Store index data along to the user requests

Author: sorry, I'm not sure about your question.
could you please show detail?

> - Store index data along to the user requests

Reviewer: Sorry I mean there is no grammar of `along to the` in english, so i couldn't understand what this sentence means.

- Search the nearest neighbor vectors of the request vector and return the search result

## Features

Vald Agent has two components, `core` and `sidecar`.

This chapter shows the characteristics of each small component.

### Core

`Core` is responsible for the main features of Vald Agent.

It uses a specific algorithm for applying features, and you can choose for your depend.

Vald provides:

- Vald Agent NGT

as core algorithm layer.

#### Vald Agent NGT

Vald Agent NGT uses [NGT](https://github.com/yahoojapan/NGT) as an algorithm.

The main functions are following:

- Insert
  - Request to insert new vectors into the NGT.
  - Requested vectors are stored in the `vqueue`.
- Search
  - Get the nearest neighbor vectors of the request vector from NGT indexes.
- Update
  - Create a request to update the specific vectors to the new vectors.
  - Requested vectors are stored in the `vqueue`.
- Delete
  - Create a request to remove the specific vectors from NGT indexes.
  - Requested vectors are stored in the `vqueue`.
- GetObject
  - Get the information of the indexed vectors.
- Exist
  - Check the specific vectors are already indexed or not.
- CreateIndex

>   - Create a new NGT index structure in memory using vectors stored in the `vqueue` and the existing NGT index structure if it exists.

Author: @kevindiu If NGT index structure already exists, the new NGT index consists of new insert vectors and it.

>   - Create a new NGT index structure in memory using vectors stored in the `vqueue` and the existing NGT index structure if it exists.

Reviewer: I think it should be more clear.
```suggestion
  - Create a new NGT index structure in memory using vectors stored in the `vqueue` and store to the existing NGT index structure.
```

>   - Create a new NGT index structure in memory using vectors stored in the `vqueue` and the existing NGT index structure if it exists.

Author: IMO, it is not correct because NGT index structure does not exist when the initial phase

- SaveIndex
  - Save metadata about NGT index information to the internal storage.

<div class="notice">
You have to control the duration of CreateIndex and SaveIndex by configuration.

These methods don’t always run when getting the request.

</div>

<div class="warning">
As you see, Vald Agent NGT can only search the nearest neighbors from the NGT index.

> You have to wait to complete CreateIndex and SaveIndex functions.

Author: I already fixed by Grammarly suggestion

</div>

This image shows the mechanism to create NGT index.

<img src="../../../assets/docs/overview/component/agent/ngt.png" />

Please refer to [Go Doc](https://pkg.go.dev/github.com/vdaas/vald@v1.3.1/pkg/agent/core/ngt/service) for other functions.

### Sidecar

`Sidecar` saves the index metadata file to external storage like Amazon S3 or Google Cloud Storage.

The main functions are:

- Backup
  - When `Agent Core` completes creating the index metadata files, `Sidecar` hooks to store these to the external storage.

<img src="../../../assets/docs/overview/component/agent/sidecar_backup.png" />

- Restore

>   - When the Vald Agent Pod reboots, the index structure is restored using external backup files.

Reviewer: ```suggestion
  - When the Vald Agent Pod restart, the index structure is restored from the external backup files.
```
